asst2/CODE/memory_finder.c

```diff
@@ -0,0 +1,438 @@
+#include <signal.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include "my_malloc.h"
+#include "global_vals.h"
+#include "memory_finder.h"
+#include "direct_mapping.h"
+#include "open_address_ht.h"
+
+/**
+ * Swaps data from page at indexA to page at indexB, updates page table and hash table to match the change
+ * @param indexA		first page index to be swapped
+ * @param indexB		second page index to be swapped
+ */
+void swap(int indexA, int indexB){
```

Comment:
are you assuming the pages will be unprotected when you do this?

Reply:
yes, we would need access to all pages to swap and stuff no ? enter and exit scheduler could maybe account for that

Reply:
this needs to be dealt with by the segfault handler. If we try to access memory we don't have access to, we will segfault .

Reply:
if all the pages are swapped and put into place before the thread runs, it won't segfault on malloc and dealloc even with swap call then, bc swap will just return right away

Reply:
we can’t do that. what if pages are on disk?

we need to fetch pages lazily

Reply:
hmm so how do we allocate new pages if we don't have access to them ?

Reply:
within malloc calls we need to unprotect then. we can probably go through the pages to find the last contiguous page allocated to a thread, and re-protect starting after the last such page.

at the end of the day, we cannot assume all of the thread memory will start out contiguous, and none of our functions can either.

Reply:
yea, i don't assume it is physically contiguous only that according to process' view it is contiguous